discover/discover.go

```diff
@@ -0,0 +1,152 @@
+package discover
+
+import (
+	"sync"
+
+	"github.com/spf13/viper"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kenjones-cisco/dapperdox/config"
+	"github.com/kenjones-cisco/dapperdox/discover/models"
+)
+
+// Discoverer represents the state of the discovery mechanism.
+type Discoverer struct {
+	services watcher
+
+	sLock sync.Mutex
+	eLock sync.Mutex
+
+	data *state
+	stop chan struct{}
+
+	vpr *viper.Viper
+
+	specs map[string][]byte
+}
+
+type state struct {
+	services    *models.ServiceMap
+	deployments *models.DeploymentMap
+}
+
+// NewDiscoverer configures a new instance of a Discoverer using Kubernetes client.
+func NewDiscoverer() (DiscoveryManager, error) {
+	log().Info("initializing new discoverer instance")
+
+	client, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+
+	// copy Viper configurations to avoid any loss of viper
+	// due to high number of goroutines in Auto-Discover process
+	vprcpy := viper.GetViper()
+
+	ctlg := newCatalog(client, catalogOptions{
+		DomainSuffix:     vprcpy.GetString(config.DiscoverySuffix),
+		WatchedNamespace: vprcpy.GetString(config.DiscoveryNamespace),
+		ResyncPeriod:     vprcpy.GetDuration(config.DiscoveryInterval),
+	})
+
+	sm := models.NewServiceMap()
+	dm := models.NewDeploymentMap()
+
+	d := &Discoverer{
+		services: ctlg,
+		data: &state{
+			services:    &sm,
+			deployments: &dm,
+		},
+		stop:  make(chan struct{}),
+		vpr:   vprcpy,
```

Comment:
is `vpr` being used anywhere?

Reply:
Should've been used inside Discover's private methods.

Reply:
removed